staging/src/k8s.io/apiserver/pkg/cel/openapi/resolver.go

```diff
@@ -0,0 +1,164 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package openapi
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/discovery"
+	"k8s.io/kube-openapi/pkg/validation/spec"
+)
+
+// SchemaResolver finds the OpenAPI schema for the given GroupVersionKind.
+// This interface uses the type defined by k8s.io/kube-openapi
+type SchemaResolver interface {
+	// ResolveSchema takes a GroupVersionKind (GVK) and returns the OpenAPI schema
+	// identified by the GVK.
+	// The function returns a non-nil error if the schema cannot be found or fail
+	// to resolve. The returned error wraps ErrSchemaNotFound if the resolution is
+	// attempted but the corresponding schema cannot be found.
+	ResolveSchema(gvk schema.GroupVersionKind) (*spec.Schema, error)
+}
+
+// ErrSchemaNotFound is wrapped and returned if the schema cannot be located
+// by the resolver.
+var ErrSchemaNotFound = fmt.Errorf("schema not found")
+
+// ClientDiscoveryResolver uses client-go discovery to resolve schemas at run time.
+type ClientDiscoveryResolver struct {
+	Discovery discovery.DiscoveryInterface
+}
+
+var _ SchemaResolver = (*ClientDiscoveryResolver)(nil)
+
+func (r *ClientDiscoveryResolver) ResolveSchema(gvk schema.GroupVersionKind) (*spec.Schema, error) {
```

Comment:
> To limit the size of this PR, it does not include the part where the admission plugin uses the library.

This looks like it is trying to prep for typed evaluation, which surprises me... I thought all the cel evaluation from admission was going to be dynamic, since we didn't know a single schema for all the types a policy would handle.

If I set up a validating admission policy that matches `apiGroups:["*"], resources: ["*"]`, will this try to resolve a schema for every type eventually?

Reply:
No. Schemas are resolved in an as-needed basis. Type checking happens when the CEL expression compiles, which in turn happens every time an expression evaluates against an object, at least as of 1.26. Type checking is optional and can be turned off in a per-expression basis. There is also plan to do an early type checking if the referred type and param can be determined.

Again this part is still being designed. I will cc you on KEP updates and such.

Reply:
making `*spec.Schema` the unifying layer to treat CRDs and built-in types consistently is a plausible goal even if the details of how validating admission will use this are still unclear

moving the existing CRD CEL validation logic from structural.Schema to *spec.Schema seems like a good first step, to avoid duplication and ensure this version still works with the CRD CEL validation

as a follow-up, as we're thinking about how validating admission would use typed evaluation, I would have questions like these:
* since CEL compilation is relatively expensive, we want to amortize compilation cost over many evaluations. can we still do that if we're compiling type-specific versions of CEL admission validation?
* how many distinct type-aware versions of a single CEL admission rule would we compile (at the limit, imagine a cluster with thousands of CRDs defined which configures an `apiGroups:["*"], apiVersions:["*"], resources: ["*"]` rule... will we compile and cache thousands of instances of the CEL program?
* admission rules can (and probably should be) created or updated in advance of the creation of the CRD they protect (or the update of the CRD to add new fields), yet that guarantees the schema will not be available yet, or will not contain the newly defined fields in the CRD. If we're using this at CRD write time to issue warnings, how do we make that a reasonable user experience?
* etc

Reply:
> moving the existing CRD CEL validation logic from structural.Schema to *spec.Schema seems like a good first step, to avoid duplication and ensure this version still works with the CRD CEL validation

As a recap, the conversion turns out to be too expensive and thus it uses an adoptor instead.

An update after the KEP finalized ([related PR](https://github.com/kubernetes/enhancements/pull/3779)). The design was simplified compared to what I had originally planned. Many of the problems considered during implementation of this library no longer exist.

> since CEL compilation is relatively expensive, we want to amortize compilation cost over many evaluations. can we still do that if we're compiling type-specific versions of CEL admission validation?
Unfortunately the evaluation remains using dynamic type, so yes, compiled expression can be cached with a lifespan same as the policy object;
> how many distinct type-aware versions of a single CEL admission rule would we compile (at the limit, imagine a cluster with thousands of CRDs defined which configures an apiGroups:["*"], apiVersions:["*"], resources: ["*"] rule... will we compile and cache thousands of instances of the CEL program?
ditto. Only one version, which is untyped.
> admission rules can (and probably should be) created or updated in advance of the creation of the CRD they protect (or the update of the CRD to add new fields), yet that guarantees the schema will not be available yet, or will not contain the newly defined fields in the CRD. If we're using this at CRD write time to issue warnings, how do we make that a reasonable user experience?
the warning comes after (1) the schema can be resolved and (2) a problem occurs. This is done as a best effort because ["Backwards-incompatible CRD changes we can treat as not our problem."](https://github.com/kubernetes/enhancements/pull/3779#discussion_r1099293474)